File: src/commands/groups/adapters.ts
import { err, ok, okBytes, textEncoder, type CommandResult } from '../../types.js';
import { errorMessage } from '../../utils.js';
import type { CommandContext, CommandSpec } from '../core.js';
import { errorCode } from '../shared/errors.js';
import { renderFetchPayload } from '../shared/json.js';
import { materializedLimitError } from '../shared/io.js';

function configuredAdapterMessage(commandName: 'search' | 'fetch'): string {
  return `${commandName}: no ${commandName} adapter configured. Pass { adapters: { ${commandName}: ... } } to createAgentCLI(...) to use this command.`;
}

function errorStatusCode(caught: unknown): number | null {
  if (!caught || typeof caught !== 'object') {
    return null;
  }

  if ('status' in caught && typeof (caught as { status?: unknown }).status === 'number') {
    return (caught as { status: number }).status;
  }
  if ('statusCode' in caught && typeof (caught as { statusCode?: unknown }).statusCode === 'number') {
    return (caught as { statusCode: number }).statusCode;
  }

  return null;
}

function isFetchNotFoundError(caught: unknown): boolean {
  const code = errorCode(caught);
  if (code === 'ENOENT' || code === 'NOT_FOUND') {
    return true;
  }

  if (errorStatusCode(caught) === 404) {
    return true;
  }

  return /\bnot found\b/i.test(errorMessage(caught));
}

async function cmdSearch(ctx: CommandContext, args: string[], stdin: Uint8Array): Promise<CommandResult> {
  if (stdin.length > 0) {
    return err('search: does not accept stdin');
  }
  if (args.length === 0) {
    return err('search: usage: search <query>');
  }
  if (!ctx.adapters.search) {
    return err(configuredAdapterMessage('search'));
  }

  try {
    const hits = await ctx.adapters.search.search(args.join(' '), 10);
    if (hits.length === 0) {
      return ok('(no results)');
    }

    const blocks = hits.map((hit, index) => {
      const lines = [`${index + 1}. ${hit.title}`];
      if (hit.source) {
        lines.push(`   source: ${hit.source}`);
      }
      if (hit.snippet) {
        lines.push(`   ${hit.snippet}`);
      }
      return lines.join('\n');
    });

    const rendered = blocks.join('\n\n');
    const renderedBytes = textEncoder.encode(rendered);
    const limitError = materializedLimitError(ctx, 'search', 'search results', renderedBytes.length);
    if (limitError !== undefined) {
      return limitError;
    }

    return okBytes(renderedBytes, 'text/plain');
  } catch (caught) {
    return err(`search: adapter error: ${errorMessage(caught)}`);
  }
}

export const search: CommandSpec = {
  name: 'search',
  summary: 'Query your configured search tool and return text results.',
  usage: 'search <query>',
  details:
    'Ranking and recall come from the configured adapter. This command formats the top results for the runtime.\nExamples:\n  search "refund timeout incident"\n  search "acme renewal risk" | head -n 5',
  handler: cmdSearch,
  acceptsStdin: false,
  minArgs: 1,
  requiresAdapter: 'search',
  conformanceArgs: ['query'],
};

async function cmdFetch(ctx: CommandContext, args: string[], stdin: Uint8Array): Promise<CommandResult> {
  if (stdin.length > 0) {
    return err('fetch: does not accept stdin');
  }
  if (args.length !== 1) {
    return err('fetch: usage: fetch <resource>');
  }
  if (!ctx.adapters.fetch) {
    return err(configuredAdapterMessage('fetch'));
  }

  try {
    const response = await ctx.adapters.fetch.fetch(args[0]!);
    return renderFetchPayload(response.payload, response.contentType, {
      commandName: 'fetch',
      subject: args[0]!,
      ...(ctx.executionPolicy.maxMaterializedBytes === undefined
        ? {}
        : { maxMaterializedBytes: ctx.executionPolicy.maxMaterializedBytes }),
    });
  } catch (caught) {
    const message = errorMessage(caught);
    if (isFetchNotFoundError(caught)) {
      return err(`fetch: resource not found: ${args[0]!}`);
    }
    return err(`fetch: adapter error: ${message}`);
  }
}

export const fetch: CommandSpec = {
  name: 'fetch',
  summary: 'Call your configured fetch tool and return text or JSON.',
  usage: 'fetch <resource>',
  details:
    'Resource identifiers and payloads come from the configured adapter. This command renders the returned text, JSON, or bytes for the runtime.\nExamples:\n  fetch order:123\n  fetch crm/customer/acme | json get owner.email',
  handler: cmdFetch,
  acceptsStdin: false,
  minArgs: 1,
  maxArgs: 1,
  requiresAdapter: 'fetch',
  conformanceArgs: ['res:1'],
};

export const adapterCommands: CommandSpec[] = [search, fetch];

File: python/src/onetool/commands/groups/adapters.py
from __future__ import annotations

from ...types import CommandResult, err, ok, ok_bytes
from ...utils import error_message
from ..core import CommandContext, CommandSpec, maybe_await
from ..shared.errors import error_code_raw
from ..shared.io import materialized_limit_error
from ..shared.json import render_fetch_payload


def _configured_adapter_message(command_name: str) -> str:
    return (
        f"{command_name}: no {command_name} adapter configured. Pass "
        f"{{ adapters: {{ {command_name}: ... }} }} to createAgentCLI(...) to use this command."
    )


def _error_status_code(caught: object) -> int | None:
    status = getattr(caught, "status", None)
    if isinstance(status, int):
        return status
    status_code = getattr(caught, "statusCode", None)
    if isinstance(status_code, int):
        return status_code
    return None


def _is_fetch_not_found_error(caught: object) -> bool:
    code = error_code_raw(caught)
    if code in {"ENOENT", "NOT_FOUND"}:
        return True
    if _error_status_code(caught) == 404:
        return True
    return "not found" in error_message(caught).lower()


async def cmd_search(ctx: CommandContext, args: list[str], stdin: bytes) -> CommandResult:
    if stdin:
        return err("search: does not accept stdin")
    if not args:
        return err("search: usage: search <query>")
    if ctx.adapters.search is None:
        return err(_configured_adapter_message("search"))

    try:
        hits = await maybe_await(ctx.adapters.search.search(" ".join(args), 10))
    except Exception as caught:
        return err(f"search: adapter error: {error_message(caught)}")

    if not hits:
        return ok("(no results)")

    blocks: list[str] = []
    for index, hit in enumerate(hits, start=1):
        lines = [f"{index}. {hit.title}"]
        if hit.source:
            lines.append(f"   source: {hit.source}")
        if hit.snippet:
            lines.append(f"   {hit.snippet}")
        blocks.append("\n".join(lines))

    rendered = "\n\n".join(blocks)
    rendered_bytes = rendered.encode("utf-8")
    limit_error = materialized_limit_error(ctx, "search", "search results", len(rendered_bytes))
    if limit_error is not None:
        return limit_error

    return ok_bytes(rendered_bytes, "text/plain")


search = CommandSpec(
    name="search",
    summary="Query your configured search tool and return text results.",
    usage="search <query>",
    details=(
        "Ranking and recall come from the configured adapter. This command formats the top results "
        "for the runtime.\n"
        "Examples:\n"
        '  search "refund timeout incident"\n'
        '  search "acme renewal risk" | head -n 5'
    ),
    handler=cmd_search,
    accepts_stdin=False,
    min_args=1,
    requires_adapter="search",
    conformance_args=("query",),
)


async def cmd_fetch(ctx: CommandContext, args: list[str], stdin: bytes) -> CommandResult:
    if stdin:
        return err("fetch: does not accept stdin")
    if len(args) != 1:
        return err("fetch: usage: fetch <resource>")
    if ctx.adapters.fetch is None:
        return err(_configured_adapter_message("fetch"))

    try:
        response = await maybe_await(ctx.adapters.fetch.fetch(args[0]))
    except Exception as caught:
        if _is_fetch_not_found_error(caught):
            return err(f"fetch: resource not found: {args[0]}")
        return err(f"fetch: adapter error: {error_message(caught)}")

    return render_fetch_payload(
        response.payload,
        response.content_type,
        command_name="fetch",
        subject=args[0],
        max_materialized_bytes=ctx.execution_policy.max_materialized_bytes,
    )


fetch = CommandSpec(
    name="fetch",
    summary="Call your configured fetch tool and return text or JSON.",
    usage="fetch <resource>",
    details=(
        "Resource identifiers and payloads come from the configured adapter. "
        "This command renders the returned text, JSON, or bytes for the runtime.\n"
        "Examples:\n"
        "  fetch order:123\n"
        "  fetch crm/customer/acme | json get owner.email"
    ),
    handler=cmd_fetch,
    accepts_stdin=False,
    min_args=1,
    max_args=1,
    requires_adapter="fetch",
    conformance_args=("res:1",),
)


adapter_commands = [search, fetch]
